lib/commons/dom/is-visible.js

```diff
@@ -179,56 +89,194 @@ function isVisible(el, screenReader, recursed) {
     return isAreaVisible(el, screenReader, recursed);
   }
 
-  // always hidden
-  if (style.getPropertyValue('display') === 'none') {
+  // different ways to hide an element
+  const hiddenMethods = [
+    displayHidden,
+    opacityHidden,
+    clipHidden,
+    scrollHidden,
+    overflowHidden,
+    visiblityHidden,
+    offscreenHidden
+  ];
+  if (
+    hiddenMethods.some(method =>
+      method({
+        el,
+        nodeName,
+        style,
+        recursed,
+        screenReader
+      })
+    )
+  ) {
     return false;
   }
 
-  // hidden from visual users
+  const parent = el.assignedSlot ? el.assignedSlot : el.parentNode;
+  let visible = false;
+  if (parent) {
+    visible = isVisible(parent, screenReader, true);
+  }
+
+  if (vNode) {
+    vNode[cacheName] = visible;
+  }
+
+  return visible;
+}
+
+/**
+ * Determine if an element is hidden by making the opacity 0
+ */
+function opacityHidden({ style, screenReader }) {
+  return !screenReader && style.getPropertyValue('opacity') === '0';
+}
+
+/**
+ * Determine if an element is hidden setting the display property
+ */
+function displayHidden({ style }) {
+  return style.getPropertyValue('display') === 'none';
+}
+
+/**
+ * Determine if an element is hidden by setting the visibility property
+ */
+function visiblityHidden({ style, recursed }) {
+  // visibility cascades so we only need to check the first element
+  return !recursed && style.getPropertyValue('visibility') === 'hidden';
+}
+
+/**
+ * Determine if an element is hidden by setting aria-hidden
+ */
+function ariaHidden({ screenReader, el, vNode }) {
+  const ariaHiddenValue = vNode
+    ? vNode.attr('aria-hidden')
+    : el.getAttribute('aria-hidden');
+  return !!screenReader && ariaHiddenValue === 'true';
+}
+
+/**
+ * Determine if an element is natively hidden
+ */
+function elementHidden({ nodeName }) {
+  return ['style', 'script', 'noscript', 'template'].includes(nodeName);
```

Comment:
I'm leaning to stick the isAreaVisible call in here, abstract that logic away from the higher isVisible method. WDYT?

Reply:
We can't. Firefox always sets the `display` style of an area element as `none`, which means it will always fail the `displayHidden` check. So we need to run the function before we call `displayHidden` and return rather than let the rest of the function continue.